diff --git a/kernel/manager/pkg_observer.c b/kernel/manager/pkg_observer.c
index 0bae3d5..137ce75 100644
--- a/kernel/manager/pkg_observer.c
+++ b/kernel/manager/pkg_observer.c
@@ -21,6 +21,28 @@ struct watch_dir {
 
 static struct fsnotify_group *g;
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
+static int ksu_handle_event(struct fsnotify_group *group, struct inode *inode,
+                            struct fsnotify_mark *inode_mark,
+                            struct fsnotify_mark *vfsmount_mark, u32 mask,
+                            const void *data, int data_type,
+                            const unsigned char *file_name, u32 cookie,
+                            struct fsnotify_iter_info *iter_info)
+{
+    if (!file_name || (mask & FS_ISDIR))
+        return 0;
+    if (!strcmp((const char *)file_name, "packages.list")) {
+        pr_info("packages.list detected: %d\n", mask);
+        track_throne(false);
+    }
+    return 0;
+}
+
+static void ksu_free_mark(struct fsnotify_mark *mark)
+{
+    kfree(mark);
+}
+#else
 static int ksu_handle_inode_event(struct fsnotify_mark *mark, u32 mask,
                                   struct inode *inode, struct inode *dir,
                                   const struct qstr *file_name, u32 cookie)
@@ -35,9 +57,15 @@ static int ksu_handle_inode_event(struct fsnotify_mark *mark, u32 mask,
     }
     return 0;
 }
+#endif
 
 static const struct fsnotify_ops ksu_ops = {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
+	.handle_event = ksu_handle_event,
+	.free_mark = ksu_free_mark,
+#else
 	.handle_inode_event = ksu_handle_inode_event,
+#endif
 };
 
 static int add_mark_on_inode(struct inode *inode, u32 mask,
@@ -52,7 +80,11 @@ static int add_mark_on_inode(struct inode *inode, u32 mask,
 	fsnotify_init_mark(m, g);
 	m->mask = mask;
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
+	if (fsnotify_add_mark(m, inode, NULL, 0)) {
+#else
 	if (fsnotify_add_inode_mark(m, inode, 0)) {
+#endif
 		fsnotify_put_mark(m);
 		return -EINVAL;
 	}
